contracts/Archetype.sol

```diff
@@ -706,4 +728,22 @@ contract Archetype is ERC721A__Initializable, ERC721AUpgradeable, OperatorFilter
   public override onlyAllowedOperator(from, royaltyEnforcementEnabled) {
       super.safeTransferFrom(from, to, tokenId, data);
   }
+
+  function _beforeTokenTransfers(address from, address to, uint256 startTokenId, uint256 quantity)
+  internal override {
+      if (pointsTokenAddress == address(0)) {
+          super._beforeTokenTransfers(from, to, startTokenId, quantity);
+      } else {
+          IPointsToken pointsToken = IPointsToken(pointsTokenAddress);
+
+          if (to != address(0) && !pointsToken.hasToken(to)) {
+              pointsToken.mintTo(to);
+          }
+
+          for (uint256 i = 0; i < quantity; i++) {
+              pointsToken.registerReceivedToken(to, startTokenId + i);
+              pointsToken.registerSentToken(from, startTokenId + i);
```

Comment:
For gas efficiency:
- Unnecessary call here on Mints where from = zero address
- I'd recommend implementing a register(Received/Sent)TokenBatch, will save a lot of gas on large mints.

Reply:
Can reduce the amount of contract calls: Here on a 5 token mint there are 12 calls to the friend contract:
https://goerli.etherscan.io/tx/0x6962a7e59f5bdda604e4c3cff6c8635dc5ad0c5d8b134bfb3b6fc3b6a03b58f0#internal

Ideally we move the logic there and make one call

Reply:
thanks for the great suggestion, consolidated the point calls on the friend contract example and it ended up being around 50% gas saving on a multi mint fuzz test.

Heres a 5 token mint on goerli with only 3 calls:
https://goerli.etherscan.io/tx/0x0356899376c5cda4382a442206ce768ca6a26899c70c10452a0a6d877980353a#internal

Heres the changes to the example: https://github.com/mini-labo/nonon-contracts/pull/4 Updated the interface in this PR as well